{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b4bace10",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import GenBank\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "353852b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_hits_df = pd.read_csv(\"/Users/jakub/Desktop/internship_helmholtz/P3/blasthits_.csv\", header = 0, delimiter = \";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "cf25cb71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>P13</th>\n",
       "      <th>P50071</th>\n",
       "      <th>P3</th>\n",
       "      <th>P4</th>\n",
       "      <th>P8</th>\n",
       "      <th>P3SEQ</th>\n",
       "      <th>P4SEQ</th>\n",
       "      <th>P8SEQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5217773-5217842</td>\n",
       "      <td>5277080-5277149</td>\n",
       "      <td>36282-36640</td>\n",
       "      <td>35111-35180</td>\n",
       "      <td>12214-12283</td>\n",
       "      <td>NODE_35</td>\n",
       "      <td>NODE_1</td>\n",
       "      <td>NODE_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1482792-1482843</td>\n",
       "      <td>5248784-5248813</td>\n",
       "      <td>22166-22235</td>\n",
       "      <td>71223-71274</td>\n",
       "      <td>76177-76228</td>\n",
       "      <td>NODE_2</td>\n",
       "      <td>NODE_24</td>\n",
       "      <td>NODE_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5017910-5017938</td>\n",
       "      <td>5486508-5486536</td>\n",
       "      <td>18870-18916</td>\n",
       "      <td>277365-277393</td>\n",
       "      <td>286437-286465</td>\n",
       "      <td>NODE_65</td>\n",
       "      <td>NODE_2</td>\n",
       "      <td>NODE_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4199733-4199780</td>\n",
       "      <td>2460746-2460797</td>\n",
       "      <td>69827-69878</td>\n",
       "      <td>26193-26220</td>\n",
       "      <td>3991-4018</td>\n",
       "      <td>NODE_4</td>\n",
       "      <td>NODE_28</td>\n",
       "      <td>NODE_67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>6218014-6218061</td>\n",
       "      <td>70317-70373</td>\n",
       "      <td>141263-141310</td>\n",
       "      <td>141263-141310</td>\n",
       "      <td>NODE_20</td>\n",
       "      <td>NODE_9</td>\n",
       "      <td>NODE_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>61304-61332</td>\n",
       "      <td>63447-63476</td>\n",
       "      <td>40550-40579</td>\n",
       "      <td>NODE_3</td>\n",
       "      <td>NODE_1</td>\n",
       "      <td>NODE_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147268-147315</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>NODE_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34852-34882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50503-50532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0              P13           P50071             P3             P4  \\\n",
       "0         NaN  5217773-5217842  5277080-5277149    36282-36640    35111-35180   \n",
       "1         NaN  1482792-1482843  5248784-5248813    22166-22235    71223-71274   \n",
       "2         NaN  5017910-5017938  5486508-5486536    18870-18916  277365-277393   \n",
       "3         NaN  4199733-4199780  2460746-2460797    69827-69878    26193-26220   \n",
       "4         NaN                X  6218014-6218061    70317-70373  141263-141310   \n",
       "5         NaN              NaN                X    61304-61332    63447-63476   \n",
       "6         NaN              NaN              NaN  147268-147315              X   \n",
       "7         NaN              NaN              NaN    34852-34882            NaN   \n",
       "8         NaN              NaN              NaN    50503-50532            NaN   \n",
       "9         NaN              NaN              NaN              X            NaN   \n",
       "\n",
       "              P8    P3SEQ    P4SEQ    P8SEQ  \n",
       "0    12214-12283  NODE_35   NODE_1  NODE_15  \n",
       "1    76177-76228   NODE_2  NODE_24  NODE_21  \n",
       "2  286437-286465  NODE_65   NODE_2   NODE_1  \n",
       "3      3991-4018   NODE_4  NODE_28  NODE_67  \n",
       "4  141263-141310  NODE_20   NODE_9   NODE_8  \n",
       "5    40550-40579   NODE_3   NODE_1  NODE_15  \n",
       "6              X   NODE_6      NaN      NaN  \n",
       "7            NaN  NODE_35      NaN      NaN  \n",
       "8            NaN   NODE_2      NaN      NaN  \n",
       "9            NaN      NaN      NaN      NaN  "
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_hits_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "54280a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_of_hits = {}\n",
    "dict_of_hits[\"P13\"] = []\n",
    "dict_of_hits[\"P50071\"] = []\n",
    "dict_of_hits[\"P3\"] = []\n",
    "dict_of_hits[\"P4\"] = []\n",
    "dict_of_hits[\"P8\"] = []\n",
    "dict_of_hits[\"P3SEQ\"] = []\n",
    "dict_of_hits[\"P4SEQ\"] = []\n",
    "dict_of_hits[\"P8SEQ\"] = []\n",
    "\n",
    "for (colname, colval) in blast_hits_df.iteritems():\n",
    "    if colname in dict_of_hits.keys():\n",
    "        dict_of_hits[colname].extend(colval.values)\n",
    "    else:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "35018f3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['P13', 'P50071', 'P3', 'P4', 'P8', 'P3SEQ', 'P4SEQ', 'P8SEQ'])\n"
     ]
    }
   ],
   "source": [
    "print(dict_of_hits.keys()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "caead2bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values([['5217773-5217842', '1482792-1482843', '5017910-5017938', '4199733-4199780', 'X', nan, nan, nan, nan, nan], ['5277080-5277149', '5248784-5248813', '5486508-5486536', '2460746-2460797', '6218014-6218061', 'X', nan, nan, nan, nan], ['36282-36640', '22166-22235', '18870-18916', '69827-69878', '70317-70373', '61304-61332', '147268-147315', '34852-34882', '50503-50532', 'X'], ['35111-35180', '71223-71274', '277365-277393', '26193-26220', '141263-141310', '63447-63476', 'X', nan, nan, nan], ['12214-12283', '76177-76228', '286437-286465', '3991-4018', '141263-141310', '40550-40579', 'X', nan, nan, nan], ['NODE_35', 'NODE_2', 'NODE_65', 'NODE_4', 'NODE_20', 'NODE_3', 'NODE_6', 'NODE_35', 'NODE_2', nan], ['NODE_1', 'NODE_24', 'NODE_2', 'NODE_28', 'NODE_9', 'NODE_1', nan, nan, nan, nan], ['NODE_15', 'NODE_21', 'NODE_1', 'NODE_67', 'NODE_8', 'NODE_15', nan, nan, nan, nan]])\n"
     ]
    }
   ],
   "source": [
    "print(dict_of_hits.values()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "9e28042e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def range_subset(range1, range2):\n",
    "    \"\"\"Whether range1 is a subset of range2.\"\"\"\n",
    "    if not range1:\n",
    "        return True  # empty range is subset of anything\n",
    "    if not range2:\n",
    "        return False  # non-empty range can't be subset of empty range\n",
    "    if len(range1) > 1 and range1.step % range2.step:\n",
    "        return False  # must have a single value or integer multiple step\n",
    "    return range1.start in range2 and range1[-1] in range2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "e565a907",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_the_hit(bacteria, gbkfile, hit_dict):\n",
    "    allCDSs = []\n",
    "    notfoundCDSs = []\n",
    "    print(\"XXXXXXXXXXXXXX\", bacteria, \"XXXXXXXXXXXXXX\")\n",
    "    for hit in hit_dict[bacteria]:\n",
    "        found_stat = False\n",
    "        if hit == \"X\":\n",
    "            break\n",
    "        else:\n",
    "            hit_start = hit.split(\"-\")[0]\n",
    "            hit_stop = hit.split(\"-\")[1]\n",
    "            print(\"The length of a hit in question: \", int(hit_stop) - int(hit_start))\n",
    "            \n",
    "            GenomeSeq = SeqIO.read(gbkfile, \"genbank\")\n",
    "\n",
    "            for i in range(len(GenomeSeq.features)):\n",
    "                current_feature = GenomeSeq.features[i]\n",
    "                if current_feature.type == 'CDS':\n",
    "                    cf_location = str(current_feature.location)\n",
    "                    gbk_start_ = cf_location.split(\":\")[0]\n",
    "                    gbk_start = int(gbk_start_.split(\"[\")[-1])\n",
    "                    gbk_stop_ = cf_location.split(\":\")[1]\n",
    "                    gbk_stop = int(gbk_stop_.split(\"]\")[0])\n",
    "                    if (gbk_start, gbk_stop, current_feature.qualifiers['product']) not in allCDSs:\n",
    "                        allCDSs.append((gbk_start, gbk_stop, current_feature.qualifiers['product']))\n",
    "                    if range_subset(range(int(hit_start), int(hit_stop)), range(int(gbk_start), int(gbk_stop))) == True:\n",
    "                        print(\"Hit found within a CDS\")\n",
    "                        found_stat = True\n",
    "                        print(hit_start, hit_stop, \"in\", gbk_start, gbk_stop, current_feature.qualifiers['product'])\n",
    "                        \n",
    "            \n",
    "        if found_stat == False:\n",
    "            notfoundCDSs.append((int(hit_start), int(hit_stop)))\n",
    "    \n",
    "    # Now for the ones outside of CDSs    \n",
    "    \n",
    "    CDSsorted_up = sorted(allCDSs)\n",
    "    for i in range(len(notfoundCDSs)):\n",
    "        possibilities_up = []\n",
    "        possibilities_down = []\n",
    "        for j in range(len(CDSsorted_up)):\n",
    "            #print(CDSsorted_up[j])\n",
    "            if notfoundCDSs[i][0] > CDSsorted_up[j][0]:\n",
    "                pass\n",
    "            else:\n",
    "                possibilities_up.extend([CDSsorted_up[j], CDSsorted_up[j+1], CDSsorted_up[j+2], CDSsorted_up[j+3], CDSsorted_up[j+5]])\n",
    "                possibilities_down.extend([CDSsorted_up[j-1], CDSsorted_up[j-2], CDSsorted_up[j-3], CDSsorted_up[j-4], CDSsorted_up[j-5]])\n",
    "                break\n",
    "\n",
    "        print(\"For this hit \",  notfoundCDSs[i], \"found upstream: \", possibilities_up)\n",
    "        print(\"For this hit \",  notfoundCDSs[i], \"found downstream: \", possibilities_down)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "48316cb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXXXXXXXXXXXX P13 XXXXXXXXXXXXXX\n",
      "The length of a hit in question:  69\n",
      "The length of a hit in question:  51\n",
      "Hit found within a CDS\n",
      "1482792 1482843 in 1479287 1483241 ['RHS repeat protein']\n",
      "The length of a hit in question:  28\n",
      "Hit found within a CDS\n",
      "5017910 5017938 in 5017746 5019720 ['methyl-accepting chemotaxis protein']\n",
      "The length of a hit in question:  47\n",
      "Hit found within a CDS\n",
      "4199733 4199780 in 4199071 4200361 ['MFS transporter']\n",
      "For this hit  (5217773, 5217842) found upstream:  [(5217883, 5218732, [\"4-(cytidine 5'-diphospho)-2-C-methyl-D-erythritol kinase\"]), (5218733, 5219351, ['lipoprotein localization protein LolB']), (5219355, 5221128, ['tetratricopeptide repeat protein']), (5221271, 5222540, ['glutamyl-tRNA reductase']), (5223641, 5224472, ['peptide chain release factor N(5)-glutamine methyltransferase'])]\n",
      "For this hit  (5217773, 5217842) found downstream:  [(5216775, 5217717, ['ribose-phosphate pyrophosphokinase']), (5216044, 5216659, ['50S ribosomal protein L25']), (5215418, 5216003, ['aminoacyl-tRNA hydrolase']), (5214277, 5215378, ['redox-regulated ATPase YchF']), (5213565, 5213871, ['addiction module antidote protein, HigA family'])]\n"
     ]
    }
   ],
   "source": [
    "check_the_hit(\"P13\", \"/Users/jakub/Desktop/internship_helmholtz/P3/final_data2/raw/genbank/p13.gbk\", dict_of_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "59a70eef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXXXXXXXXXXXX P50071 XXXXXXXXXXXXXX\n",
      "The length of a hit in question:  69\n",
      "The length of a hit in question:  29\n",
      "The length of a hit in question:  28\n",
      "Hit found within a CDS\n",
      "5486508 5486536 in 5484725 5486699 ['chemotaxis protein']\n",
      "The length of a hit in question:  51\n",
      "Hit found within a CDS\n",
      "2460746 2460797 in 2460347 2464301 ['Rhs family protein']\n",
      "The length of a hit in question:  47\n",
      "Hit found within a CDS\n",
      "6218014 6218061 in 6217432 6218722 ['citrate-proton symporter']\n",
      "For this hit  (5277080, 5277149) found upstream:  [(5277204, 5278146, ['ribose-phosphate pyrophosphokinase']), (5278262, 5278877, ['50S ribosomal protein L25']), (5278918, 5279503, ['peptidyl-tRNA hydrolase']), (5279543, 5280644, ['GTP-binding protein']), (5281972, 5284201, ['TonB-dependent receptor'])]\n",
      "For this hit  (5277080, 5277149) found downstream:  [(5276189, 5277038, ['kinase']), (5275570, 5276188, ['membrane protein']), (5273838, 5275566, ['hypothetical protein']), (5272381, 5273650, ['glutamyl-tRNA reductase']), (5271281, 5272364, ['peptide chain release factor 1'])]\n",
      "For this hit  (5248784, 5248813) found upstream:  [(5248821, 5249235, ['3-demethylubiquinone-9 3-methyltransferase']), (5249503, 5249794, ['aminopeptidase']), (5249876, 5250362, ['hypothetical protein']), (5250429, 5250903, ['cupin']), (5251629, 5252268, ['uracil phosphoribosyltransferase'])]\n",
      "For this hit  (5248784, 5248813) found downstream:  [(5247248, 5248772, ['malate:quinone oxidoreductase']), (5245819, 5246407, ['lipoprotein']), (5245555, 5245762, ['hypothetical protein']), (5244975, 5245272, ['lipoprotein']), (5243694, 5244855, ['glycerol acyltransferase'])]\n"
     ]
    }
   ],
   "source": [
    "check_the_hit(\"P50071\", \"/Users/jakub/Desktop/internship_helmholtz/P3/final_data2/raw/genbank/p50071.gbk\", dict_of_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "606a361c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_the_hit_assembly(bacteria, node_col_id, gbkfile, hit_dict):\n",
    "    allCDSs = []\n",
    "    notfoundCDSs = []\n",
    "    print(\"XXXXXXXXXXXXXX\", bacteria, \"XXXXXXXXXXXXXX\")\n",
    "    for h in range(len(hit_dict[bacteria])):\n",
    "        found_stat = False\n",
    "        if hit_dict[bacteria][h] == \"X\":\n",
    "            break\n",
    "        else:\n",
    "            hit_start = hit_dict[bacteria][h].split(\"-\")[0]\n",
    "            hit_stop = hit_dict[bacteria][h].split(\"-\")[1]\n",
    "            print(\"The length of a hit in question: \", int(hit_stop) - int(hit_start)) \n",
    "            for record in SeqIO.parse(gbkfile, \"genbank\"):\n",
    "                for i in range(len(record.features)):\n",
    "                    current_feature = record.features[i]\n",
    "                    if current_feature.type == 'CDS':\n",
    "                        cf_location = str(current_feature.location)\n",
    "                        gbk_start_ = cf_location.split(\":\")[0]\n",
    "                        if \"<\" in gbk_start_:\n",
    "                            gbk_start_ = gbk_start_.split('<')[-1]\n",
    "                        gbk_start = int(gbk_start_.split(\"[\")[-1])\n",
    "                        gbk_stop_ = cf_location.split(\":\")[1]\n",
    "                        if \">\" in gbk_stop_:\n",
    "                            gbk_stop_ = gbk_stop_.split('>')[-1]\n",
    "                        gbk_stop = int(gbk_stop_.split(\"]\")[0])\n",
    "                        if (gbk_start, gbk_stop, current_feature.qualifiers['product'], record.id) not in allCDSs:\n",
    "                            allCDSs.append((gbk_start, gbk_stop, current_feature.qualifiers['product'], record.id))\n",
    "                        if range_subset(range(int(hit_start), int(hit_stop)), range(int(gbk_start), int(gbk_stop))) == True:\n",
    "                            if record.id == hit_dict[node_col_id][h]:                                \n",
    "                                print(\"Hit found within a CDS\")\n",
    "                                found_stat = True\n",
    "                                print(hit_start, hit_stop, \"in\", gbk_start, gbk_stop, current_feature.qualifiers['product'])\n",
    "                        \n",
    "            \n",
    "        if found_stat == False:\n",
    "            notfoundCDSs.append((int(hit_start), int(hit_stop), hit_dict[node_col_id][h]))\n",
    "    \n",
    "    #Now for the ones outside of CDSs     \n",
    "    \n",
    "    CDSsorted_up = sorted(allCDSs)\n",
    "    for i in range(len(notfoundCDSs)):\n",
    "        possibilities_up = []\n",
    "        possibilities_down = []\n",
    "        for j in range(len(CDSsorted_up)):\n",
    "            if notfoundCDSs[i][-1] == CDSsorted_up[j][-1]:\n",
    "                if notfoundCDSs[i][0] > CDSsorted_up[j][0]:\n",
    "                    pass\n",
    "                else:\n",
    "                    possibilities_up.extend([CDSsorted_up[j], CDSsorted_up[j+1], CDSsorted_up[j+2], CDSsorted_up[j+3], CDSsorted_up[j+5]])\n",
    "                    possibilities_down.extend([CDSsorted_up[j-1], CDSsorted_up[j-2], CDSsorted_up[j-3], CDSsorted_up[j-4], CDSsorted_up[j-5]])\n",
    "                    break\n",
    "\n",
    "        print(\"For this hit \",  notfoundCDSs[i], \"found upstream: \", possibilities_up)\n",
    "        print(\"For this hit \",  notfoundCDSs[i], \"found downstream: \", possibilities_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "807c2d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXXXXXXXXXXXX P3 XXXXXXXXXXXXXX\n",
      "The length of a hit in question:  358\n",
      "Hit found within a CDS\n",
      "36282 36640 in 35711 36683 ['Major capsid protein 13-like']\n",
      "The length of a hit in question:  69\n",
      "The length of a hit in question:  46\n",
      "Hit found within a CDS\n",
      "18870 18916 in 18808 22654 ['cobaltochelatase subunit CobN']\n",
      "The length of a hit in question:  51\n",
      "Hit found within a CDS\n",
      "69827 69878 in 66322 70276 ['Toxin protein Tse5']\n",
      "The length of a hit in question:  56\n",
      "The length of a hit in question:  28\n",
      "Hit found within a CDS\n",
      "61304 61332 in 59521 61495 ['Methyl-accepting chemotaxis protein CtpL']\n",
      "The length of a hit in question:  47\n",
      "Hit found within a CDS\n",
      "147268 147315 in 146606 147896 ['citrate-proton symporter']\n",
      "The length of a hit in question:  30\n",
      "The length of a hit in question:  29\n",
      "For this hit  (22166, 22235, 'NODE_2') found upstream:  [(22276, 23125, [\"4-(cytidine 5'-diphospho)-2-C-methyl-D-erythritol kinase\"], 'NODE_2'), (22285, 22774, ['DUF1833 domain-containing protein'], 'NODE_53'), (22294, 23227, ['Integrating conjugative element protein'], 'NODE_25'), (22295, 23192, ['methylisocitrate lyase'], 'NODE_21'), (22300, 23245, ['1-phosphofructokinase'], 'NODE_38')]\n",
      "For this hit  (22166, 22235, 'NODE_2') found downstream:  [(22271, 22886, ['TetR family transcriptional regulator'], 'NODE_14'), (22268, 22697, ['cobalamin biosynthesis protein'], 'NODE_52'), (22265, 22538, ['DNA-binding protein HU-alpha'], 'NODE_20'), (22248, 22755, ['transcriptional regulator'], 'NODE_66'), (22225, 24811, ['glucan biosynthesis glucosyltransferase H'], 'NODE_26')]\n",
      "For this hit  (70317, 70373, 'NODE_20') found upstream:  [(70466, 72395, ['acetyl-CoA hydrolase'], 'NODE_20'), (70586, 71870, ['C4-dicarboxylate TRAP transporter large permease protein DctM'], 'NODE_1'), (70615, 73252, ['phosphoenolpyruvate carboxylase'], 'NODE_11'), (70630, 71173, ['2-amino-4-hydroxy-6-hydroxymethyldihydropteridine diphosphokinase'], 'NODE_10'), (70691, 71456, ['N(G),N(G)-dimethylarginine dimethylaminohydrolase'], 'NODE_24')]\n",
      "For this hit  (70317, 70373, 'NODE_20') found downstream:  [(70434, 73566, ['multidrug efflux RND transporter permease subunit MuxB'], 'NODE_14'), (70404, 72252, ['SBP-bac-5 domain-containing protein'], 'NODE_27'), (70391, 70715, ['DNA-binding transcriptional regulator Fis'], 'NODE_3'), (70325, 72395, ['Type VI secretion system spike protein VgrG1c'], 'NODE_4'), (70319, 70796, ['transcriptional regulator'], 'NODE_7')]\n",
      "For this hit  (34852, 34882, 'NODE_35') found upstream:  [(34958, 35702, ['DUF6651 domain-containing protein'], 'NODE_35'), (34981, 35716, ['Integral membrane protein TerC family'], 'NODE_57'), (35024, 35840, ['phosphonate C-P lyase system protein PhnK'], 'NODE_5'), (35068, 35575, ['diacylglycerol kinase'], 'NODE_23'), (35079, 36006, ['ribokinase'], 'NODE_49')]\n",
      "For this hit  (34852, 34882, 'NODE_35') found downstream:  [(34902, 35781, ['LysR family transcriptional regulator'], 'NODE_14'), (34873, 35935, ['Putrescine/cadaverine-binding protein'], 'NODE_22'), (34868, 35108, ['Lipoprotein'], 'NODE_53'), (34834, 37207, ['Type VI secretion system spike protein VgrG5'], 'NODE_26'), (34790, 36764, ['Cytochrome c-type biogenesis protein CcmF'], 'NODE_31')]\n",
      "For this hit  (50503, 50532, 'NODE_2') found upstream:  [(50543, 52067, ['malate dehydrogenase (quinone)'], 'NODE_2'), (50548, 51661, ['2-oxo acid dehydrogenase subunit E2'], 'NODE_44'), (50593, 51418, ['AraC family transcriptional regulator'], 'NODE_16'), (50648, 51389, ['flagellar motor protein'], 'NODE_31'), (50755, 51157, ['GFA family protein'], 'NODE_45')]\n",
      "For this hit  (50503, 50532, 'NODE_2') found downstream:  [(50500, 51862, ['aminodeoxychorismate synthase component I'], 'NODE_19'), (50481, 52638, ['TonB-dependent siderophore receptor'], 'NODE_41'), (50474, 51536, ['gluconolactonase'], 'NODE_42'), (50440, 51238, ['phosphate ABC transporter substrate-binding protein'], 'NODE_26'), (50395, 51874, ['long-chain fatty acid--CoA ligase'], 'NODE_24')]\n"
     ]
    }
   ],
   "source": [
    "check_the_hit_assembly(\"P3\", \"P3SEQ\", \"/Users/jakub/Desktop/internship_helmholtz/P3/final_data2/raw/genbank/p3.gbk\", dict_of_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "e585289b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXXXXXXXXXXXX P4 XXXXXXXXXXXXXX\n",
      "The length of a hit in question:  69\n",
      "The length of a hit in question:  51\n",
      "Hit found within a CDS\n",
      "71223 71274 in 67718 71672 ['Toxin protein Tse5']\n",
      "The length of a hit in question:  28\n",
      "Hit found within a CDS\n",
      "277365 277393 in 277201 279175 ['Methyl-accepting chemotaxis protein CtpL']\n",
      "The length of a hit in question:  27\n",
      "Hit found within a CDS\n",
      "26193 26220 in 26062 27730 ['DNA helicase']\n",
      "The length of a hit in question:  47\n",
      "Hit found within a CDS\n",
      "141263 141310 in 140601 141891 ['citrate-proton symporter']\n",
      "The length of a hit in question:  29\n",
      "For this hit  (35111, 35180, 'NODE_1') found upstream:  [(35221, 36070, [\"4-(cytidine 5'-diphospho)-2-C-methyl-D-erythritol kinase\"], 'NODE_1'), (35224, 35758, ['translation initiation factor IF-3'], 'NODE_28'), (35260, 36121, ['carboxylate/amino acid/amine transporter'], 'NODE_45'), (35275, 36724, ['two-component system sensor histidine kinase GtrS'], 'NODE_50'), (35327, 35531, ['YscE family type III secretion system co-chaperone PscE'], 'NODE_4')]\n",
      "For this hit  (35111, 35180, 'NODE_1') found downstream:  [(35207, 36011, ['hypothetical protein'], 'NODE_6'), (35169, 35718, [\"RNA 2',3'-cyclic phosphodiesterase\"], 'NODE_52'), (35088, 36234, ['FMNH2-dependent alkanesulfonate monooxygenase'], 'NODE_31'), (35086, 35857, ['hypothetical protein'], 'NODE_26'), (35085, 35688, ['OMP-b-brl domain-containing protein'], 'NODE_16')]\n",
      "For this hit  (63447, 63476, 'NODE_1') found upstream:  [(63489, 65013, ['malate dehydrogenase (quinone)'], 'NODE_1'), (63568, 64210, ['3-beta hydroxysteroid dehydrogenase'], 'NODE_20'), (63577, 64324, ['phage recombination protein Bet'], 'NODE_18'), (63621, 63960, ['PBCV-basic-adap domain-containing protein'], 'NODE_7'), (63638, 64715, ['3-deoxy-7-phosphoheptulonate synthase'], 'NODE_4')]\n",
      "For this hit  (63447, 63476, 'NODE_1') found downstream:  [(63415, 64585, ['amidohydrolase'], 'NODE_22'), (63401, 64538, ['flagellar biosynthetic protein FlhB'], 'NODE_21'), (63392, 64646, ['cell division protein FtsA'], 'NODE_13'), (63352, 64531, ['protein disulfide reductase'], 'NODE_35'), (63338, 64109, ['Imidazole glycerol phosphate synthase subunit HisF'], 'NODE_8')]\n"
     ]
    }
   ],
   "source": [
    "check_the_hit_assembly(\"P4\", \"P4SEQ\",\"/Users/jakub/Desktop/internship_helmholtz/P3/final_data2/raw/genbank/p4.gbk\", dict_of_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "f498113e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXXXXXXXXXXXX P8 XXXXXXXXXXXXXX\n",
      "The length of a hit in question:  69\n",
      "The length of a hit in question:  51\n",
      "Hit found within a CDS\n",
      "76177 76228 in 72672 76626 ['Toxin protein Tse5']\n",
      "The length of a hit in question:  28\n",
      "Hit found within a CDS\n",
      "286437 286465 in 286273 288247 ['Methyl-accepting chemotaxis protein CtpL']\n",
      "The length of a hit in question:  27\n",
      "Hit found within a CDS\n",
      "3991 4018 in 2480 4148 ['DNA helicase']\n",
      "The length of a hit in question:  47\n",
      "Hit found within a CDS\n",
      "141263 141310 in 140601 141891 ['citrate-proton symporter']\n",
      "The length of a hit in question:  29\n",
      "For this hit  (12214, 12283, 'NODE_15') found upstream:  [(12324, 13173, [\"4-(cytidine 5'-diphospho)-2-C-methyl-D-erythritol kinase\"], 'NODE_15'), (12329, 12950, ['DUF2170 domain-containing protein'], 'NODE_34'), (12344, 13046, ['nuclease'], 'NODE_50'), (12350, 13478, ['FAD-dependent catabolic D-arginine dehydrogenase DauA'], 'NODE_53'), (12405, 12912, ['peptidoglycan-associated lipoprotein Pal'], 'NODE_61')]\n",
      "For this hit  (12214, 12283, 'NODE_15') found downstream:  [(12319, 12580, ['Chromate resistance protein'], 'NODE_45'), (12283, 12901, ['glycoprotein'], 'NODE_57'), (12270, 13044, ['electron transfer flavoprotein subunit beta'], 'NODE_43'), (12260, 13061, ['amino acid ABC transporter substrate-binding protein'], 'NODE_1'), (12259, 13177, ['LysR family transcriptional regulator'], 'NODE_19')]\n",
      "For this hit  (40550, 40579, 'NODE_15') found upstream:  [(40592, 42116, ['malate dehydrogenase (quinone)'], 'NODE_15'), (40658, 41768, ['S-(hydroxymethyl)glutathione dehydrogenase/class III alcohol dehydrogenase'], 'NODE_32'), (40695, 41778, ['secretion system protein F'], 'NODE_30'), (40696, 41392, ['Mks condensin complex protein MksE'], 'NODE_45'), (40760, 41747, ['NADH:quinone reductase'], 'NODE_14')]\n",
      "For this hit  (40550, 40579, 'NODE_15') found downstream:  [(40589, 41810, ['amidohydrolase'], 'NODE_13'), (40479, 40599, ['Regulator'], 'NODE_32'), (40460, 41213, ['hypothetical protein'], 'NODE_26'), (40404, 41694, ['histidine--tRNA ligase'], 'NODE_22'), (40394, 40799, ['chromosome partitioning protein ParA'], 'NODE_33')]\n"
     ]
    }
   ],
   "source": [
    "check_the_hit_assembly(\"P8\", \"P8SEQ\",\"/Users/jakub/Desktop/internship_helmholtz/P3/final_data2/raw/genbank/p8.gbk\", dict_of_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f2a680",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
